### PR TITLE
Add support for D-Bus-based yggdrasil IPC

### PR DESCRIPTION
**Pull request by subpop**
Add parallel support for communicating with yggd over D-Bus. In order to support both gRPC and D-Bus, the gRPC-based yggdrasil package is included in the source tree of this project and a go module replace directive is used to replace the external package lookup with a local one.

When the program starts, it decides to use gRPC only if the YGG_SOCKET_ADDR variable is present in the environment. If that variable is absent, the worker connects to D-Bus using the upstream yggdrasil "worker" package API.

### CODE REVIEW

**Review by ekohl — COMMENTED** [review]
Looks like this is vendoring in a complete yggdrasil version. Is that intended?

To make testing easier, I started to add Packit in https://github.com/theforeman/foreman_ygg_worker/pull/20.

**Comment by subpop** [comment]
Yes, this is deliberate. This is how one worker can support both the D-Bus and the gRPC versions of yggdrasil. You'll notice one of the commits pulls the yggdrasil 0.2.x source code into the repository and renames it to "_v0".

**Comment by adamruzicka** [comment]
Could we add yggdrasil as a git submodule instead of carrying a copy of it?

**Comment by subpop** [comment]
> Could we add yggdrasil as a git submodule instead of carrying a copy of it?

Good idea. I'll refactor to do that.

**Comment by subpop** [comment]
See RedHatInsights/yggdrasil#250.

**Comment by subpop** [comment]
I updated the PR to use a submodule. It turns out I was able to do it with a combination of submodules and go module replace directives. The changes I thought were required upstream (RedHatInsights/yggdrasil#250) were not needed for this particular implementation.

**Comment by ekohl** [comment]
I kicked off CI, but it's not happy. For the future I also relaxed the repo's GitHub Actions settings from `Require approval for first-time contributors` to `Require approval for first-time contributors who are new to GitHub` which should allow you to iterate easier. And to be honest, because I'm lazy and don't want to approve every CI run after you push ;)

**Comment by subpop** [comment]
Tests are passing now.

**Comment by ekohl** [comment]
I'm looking at @adamruzicka for the real code. I only look at the build system/CI side of things.

@subpop could you rebase on top of main so that we get packit builds as well?

File: Makefile
Comment (ekohl) [inline Makefile:20]: `LIBEXECDIR` already starts with `/` so this is redundant
```suggestion
	install -D -m 755 build/foreman_worker $(DESTDIR)$(LIBEXECDIR)/foreman_worker
```

File: Makefile
Comment (ekohl) [inline Makefile:55]: Any particular reason you write to the `.tmp` file? Isn't this sufficient?
```suggestion
		$< > $@
```

File: data/com.redhat.Yggdrasil1.Worker1.foreman.service.in
Comment (ekohl) [inline data/com.redhat.Yggdrasil1.Worker1.foreman.service.in:8]: It's confusing to me to use `worker_user` as a variable to determine the group. I also wonder what's the benefit of running as root but with a different group.

File: data/com.redhat.Yggdrasil1.Worker1.foreman.service.in
Comment (ekohl) [inline data/com.redhat.Yggdrasil1.Worker1.foreman.service.in:2]: I don't see a short description for this repo on GH, but the first line in README reads:
> This is a worker service for yggdrasil that can act as pull client for Foreman Remote Execution.

So something like this perhaps?
```suggestion
Description=An yggdrasil worker service to act as a pull client for Foreman Remote Execution
```

@adamruzicka is that a good summary we should also use for the GH repo?

**Review by ekohl — COMMENTED** [review]
Nice, having working RPMs should make testing easier.

File: Makefile
Comment (adamruzicka) [inline Makefile:23]: Huh, interesting. I would expect the rpm build to fail due to additional files. But the build seems to pass and those files are note present in the resulting rpm.

File: Makefile
Comment (subpop) [reply Makefile:55]: No, just a holdover from old Makefile patterns.

File: data/com.redhat.Yggdrasil1.Worker1.foreman.service.in
Comment (subpop) [reply data/com.redhat.Yggdrasil1.Worker1.foreman.service.in:2]: Oops. This was a copy-paste error.

File: data/com.redhat.Yggdrasil1.Worker1.foreman.service.in
Comment (subpop) [reply data/com.redhat.Yggdrasil1.Worker1.foreman.service.in:8]: This is a side-effect of the yggd D-Bus architecture. In order to own bus names
under the namespace "com.redhat.Yggdrasil1.Worker1", a process must have the
permission to do so. yggdrasil installs a D-Bus policy that restricts processes
that are members of a group (defined at compile-time).
https://github.com/RedHatInsights/yggdrasil/commit/16fec576da5b4d436abdc5c3cad7b2a33cca637

File: Makefile
Comment (ekohl) [inline Makefile:51]: At first I thought this should be `build/data` but reading https://www.gnu.org/software/make/manual/html_node/Automatic-Variables.html you can automatically get the directory name of the target:
```suggestion
	mkdir -p $(@D)
```

File: data/com.redhat.Yggdrasil1.Worker1.foreman.service.in
Comment (ekohl) [reply data/com.redhat.Yggdrasil1.Worker1.foreman.service.in:8]: Can we at least rename it to `@worker_group@` to emphasize what it's used for?

File: data/com.redhat.Yggdrasil1.Worker1.foreman.service.in
Comment (ekohl) [inline data/com.redhat.Yggdrasil1.Worker1.foreman.service.in:2]:
```suggestion
Description=Foreman remote execution yggdrasil worker service
```

File: data/com.redhat.Yggdrasil1.Worker1.foreman.service.in
Comment (subpop) [reply data/com.redhat.Yggdrasil1.Worker1.foreman.service.in:8]: Yea, good point.

**Comment by ekohl** [comment]
From a build perspective :+1: but I'm not qualified to review the Go code.

**Review by adamruzicka — APPROVED** [review]
None of the things I complained in the previous comment can really be addressed in this PR, so I take that back.

The change proposed here look sane and I was able to get the entire flow to work end to end with yggdrasil-0.2.1 and 0.4.4, let's get this in.

**Comment by adamruzicka** [comment]
Thank you @subpop & @ekohl !

**Pull request merged**
